File: lua/puzzle7-1.lua
require "utils"

local function get_type(hand)
	hand = {unpack(hand)}
	table.sort(hand)

	local groups = {}
	local cur = table.remove(hand)
	local n = 1
	while hand[1] do
		local c = table.remove(hand)
		if c == cur then
			n = n + 1
		else
			table.insert(groups, n)
			cur = c
			n = 1
		end
	end
	table.insert(groups, n)

	local len = #groups
	if len == 1 then
		return 7 -- 5 of a kind
	elseif len == 2 then
		if groups[1] == 4 or groups[2] == 4 then
			return 6 -- 4 of a kind
		end
		return 5 -- full house
	elseif len == 3 then
		if groups[1] == 2 or groups[2] == 2 then
			return 3 -- 2 pair
		end
		return 4 -- 3 of a kind
	elseif len == 4 then
		return 2 -- 1 pair
	else
		return 1 -- high card
	end
end

local labels = {"2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"}
local strength = {}
for i, v in ipairs(labels) do
	strength[v] = i
end

local function cmp(hand1, hand2)
	if hand1 == hand2 then return false end -- same hand
	if hand1._type == hand2._type then
		local c1, c2 = hand1.cards, hand2.cards
		for i = 1, #c1 do
			local s1, s2 = strength[c1[i]], strength[c2[i]]
			if s1 ~= s2 then return s1 < s2 end
		end
		assert(false) -- duplicates
	else
		return hand1._type < hand2._type
	end
end

local hands = {}
for line in io.lines() do
	local chars, bid = unpack(string.split(line))
	assert(chars, bid)
	local cards = {}
	for i = 1, #chars do
		table.insert(cards, string.sub(chars, i, i))
	end
	hand = {cards=cards, bid=bid + 0, _type=get_type(cards)}
	table.insert(hands, hand)
end

table.sort(hands, cmp)
local sum = 0
for i, hand in ipairs(hands) do
	sum = sum + i * hand.bid
end
print(sum)

File: c++/puzzle7-1.cpp
#include <iostream>
#include <string>
#include <charconv>
#include <cassert>
#include <vector>
#include <algorithm>
#include <unordered_map>

enum class Type { HIGH, ONE, TWO, THREE, HOUSE, FOUR, FIVE };

static std::unordered_map<char, unsigned> strength;

struct Hand {
    std::string cards;
    unsigned long bid;
    Type type;

    bool operator<(const Hand& other) const
    {
        if (type != other.type)
            return type < other.type;
        else {
            assert(cards.size() == other.cards.size());
            for (decltype(cards)::size_type i = 0; i < cards.size(); ++i) {
                auto card = cards[i];
                auto oth = other.cards[i];
                if (card != oth)
                    return strength.at(card) < strength.at(oth);
            }
            assert(0);
        }
    }
};

Type get_type(std::string cards)
{
    std::sort(cards.begin(), cards.end());
    std::vector<unsigned> groups;
    auto cur = cards[0];
    unsigned group = 1;
    for (auto card : cards.substr(1)) {
        if (card == cur)
            ++group;
        else {
            groups.push_back(group);
            cur = card;
            group = 1;
        }
    }
    groups.push_back(group);

    auto n = groups.size();
    assert(n > 0 && n <= 5);
    if (n == 1)
        return Type::FIVE;
    else if (n == 2)
        return groups[0] * groups[1] == 4 ? Type::FOUR : Type::HOUSE;
    else if (n == 3)
        return groups[0] * groups[1] % 2 ? Type::THREE : Type::TWO;
    else if (n == 4)
        return Type::ONE;
    else
        return Type::HIGH;
}

std::vector<Hand> parse_hands()
{
    std::vector<Hand> hands;
    for (std::string line; getline(std::cin, line);) {
        auto pos = line.find(' ');
        assert(pos == 5);
        std::string cards = line.substr(0, pos);

        unsigned long bid = 0;
        auto res = std::from_chars(line.data() + pos + 1,
            line.data() + line.size(), bid);
        assert(res.ec == std::errc{} && bid);

        auto type = get_type(cards);
        hands.push_back({std::move(cards), bid, type});
    }
    return hands;
}

int main()
{
    std::string labels = "AKQJT98765432";
    for (std::string::size_type i = 0; i < labels.size(); ++i)
        strength[labels[i]] = labels.size() - i;

    auto hands = parse_hands();
    std::sort(hands.begin(), hands.end());

    unsigned long sum = 0;
    for (decltype(hands)::size_type i = 0; i < hands.size(); ++i)
        sum += (i + 1) * hands[i].bid;
    std::cout << sum << '\n';
    return 0;
}
